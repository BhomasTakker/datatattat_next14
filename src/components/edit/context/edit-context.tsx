import { ReactNode, createContext } from "react";

type EditState = {
	submitHandler: (data: any) => void;
	submitDraftHandler: (data: any) => void;
	submitDebugHandler?: (data: any) => void;
	saveAsTemplateHandler?: (data: any) => void;
};

type EditInterface = {};

const initialState: EditState & EditInterface = {
	submitHandler: () => {
		// Default implementation
	},
	submitDraftHandler: () => {
		// Default implementation
	},
	submitDebugHandler: () => {
		// Default implementation
	},
	saveAsTemplateHandler: () => {
		// Default implementation
	},
};

export const EditContextProvider = ({
	children,
	value,
}: {
	children: ReactNode;
	value: EditState;
}) => {
	return (
		<EditContext.Provider value={{ ...value }}>{children}</EditContext.Provider>
	);
};

export const EditContext = createContext({ ...initialState });
