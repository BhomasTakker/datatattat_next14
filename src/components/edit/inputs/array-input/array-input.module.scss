@use "../../../../scss/abstracts/units" as *;
@use "../../../../scss/abstracts/ui/" as *;
@use "../../../../scss/abstracts/colours/" as *;

.root {
	display: flex;
	flex-direction: column;
	gap: $spacingMd;

	// border: 1px solid #ccc;
	box-shadow: 0 0 $spacingXxs $shadowLight3;

	margin-top: $spacingMd;
	padding: $spacingMd;

	.title {
	}

	.input {
		display: flex;
		flex-direction: column;
		gap: $spacingXs;

		.icons {
			display: flex;
			gap: $spacingXs;
			align-items: center;
			align-self: end;
		}

		@media screen and (max-width: $screenWidthMd) {
			flex-direction: column-reverse;
		}
	}
	.input:not(:last-of-type) {
		border-bottom: $divider;
		padding-bottom: $spacingMd;
	}
}
