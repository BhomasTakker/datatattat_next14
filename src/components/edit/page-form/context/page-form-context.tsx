import { savePage } from "@/actions/edit/update-page";
import { IPage } from "@/types/page";
import { createContext, ReactNode, useState } from "react";
import { FieldValues, UseFormReturn } from "react-hook-form";

// get from config
const ROUTE = "route";
const CREATOR = "creator";
const CONTENT = "content";
const META = "meta";
const PROFILE = "profile";

const formIds = {
	creator: CREATOR,
	content: CONTENT,
	meta: META,
	profile: PROFILE,
	route: ROUTE,
};

type PageFormContextProviderProps = {
	children: ReactNode;
	page: IPage;
	methods: UseFormReturn<FieldValues, unknown, undefined>;
};

type PageFormContextInitialState = {
	page: IPage | null;
	pageState: string | undefined;
	submitHandler: (e?: React.BaseSyntheticEvent) => Promise<void>;
	formIds: typeof formIds;
	methods: UseFormReturn<FieldValues, unknown, undefined>;
};

const initialState: PageFormContextInitialState = {
	page: null,
	pageState: undefined,
	submitHandler: async () => {},
	formIds,
	methods: {} as UseFormReturn<FieldValues, unknown, undefined>,
};

export const PageFormContextProvider = ({
	children,
	page,
	methods,
}: PageFormContextProviderProps) => {
	const [pageState, setPageState] = useState<string | undefined>(undefined);
	const { handleSubmit } = methods;

	const { route } = page;

	const submitHandler = handleSubmit(async (data) => {
		// This should be done on the server...
		// Take what's changed and merge with the page object
		const pageData = { ...page, ...data };

		// get meta value
		// get profile value
		// get content value
		// get creator_id value
		// create page object
		// save page object
		// revalidate the page path
		// set page state
		const res = await savePage(route, pageData);

		setPageState(res.message);
	});

	return (
		<PageFormContext.Provider
			value={{ page, pageState, submitHandler, formIds, methods }}
		>
			{children}
		</PageFormContext.Provider>
	);
};

export const PageFormContext = createContext({ ...initialState });
