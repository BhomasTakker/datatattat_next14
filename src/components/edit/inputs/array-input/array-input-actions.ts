import { GenericInput } from "@/types/edit/inputs/inputs";
import { Dispatch, SetStateAction } from "react";
import {
	FieldValues,
	UseFormGetValues,
	UseFormSetValue,
} from "react-hook-form";

type Direction = "up" | "down";

type AddHOCProps = {
	setValue: UseFormSetValue<FieldValues>;
	getValues: UseFormGetValues<FieldValues>;
	setInputList: Dispatch<SetStateAction<GenericInput[]>>;
	id: string;
	inputId: string;
	createObject: boolean;
};

type DeleteHOCProps = {
	inputs: GenericInput[];
	setValue: UseFormSetValue<FieldValues>;
	setInputList: Dispatch<SetStateAction<GenericInput[]>>;
	id: string;
};

type MoveHOCProps = {
	inputs: GenericInput[];
	setValue: UseFormSetValue<FieldValues>;
	setInputList: Dispatch<SetStateAction<GenericInput[]>>;
	id: string;
};

export const initialise = () => {};

// Not sure about create object?
// The name at least is poor
export const add =
	({
		setValue,
		getValues,
		setInputList,
		id,
		inputId,
		createObject,
	}: AddHOCProps) =>
	() => {
		const inputList: GenericInput[] = getValues(id) || [];
		// we set value here to give us something to read in InputList
		// there may be a better way to do this
		if (createObject) {
			setValue(id, [...inputList, { [`${inputId}`]: "" }]);
		} else {
			setValue(id, [...inputList, ""]);
		}

		setInputList(getValues(id));
	};

export const onDelete =
	({ inputs, setValue, setInputList, id }: DeleteHOCProps) =>
	(index: number) => {
		const newArray = inputs.filter((_, i) => index !== i);
		setValue(id, newArray);
		setInputList(newArray);
	};

export const move =
	({ inputs, setValue, setInputList, id }: MoveHOCProps) =>
	(index: number, direction: Direction) => {
		if (index === 0 && direction === "up") return;
		if (index === inputs.length - 1 && direction === "down") return;

		// technically we should perhaps setValues
		// WITHIN the useState setter function
		// for both move and delete
		const newIndex = direction === "up" ? index - 1 : index + 1;
		const item = inputs[index];
		const removed = inputs.toSpliced(index, 1);
		const moved = removed.toSpliced(newIndex, 0, item);
		setValue(id, moved);
		setInputList(moved);
	};
