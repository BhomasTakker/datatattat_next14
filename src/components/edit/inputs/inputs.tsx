import { FC } from "react";
import { InputTitle } from "./title/input-title";
import { TextInput } from "./text-input/text-input";
import { ShowInput } from "./show-input/show-input";
import { ArrayInput } from "./array-input/array-input";
import { InputList } from "./input-list/input-list";
import { SelectInput } from "./select-input/select-input";
import { ObjectSelect } from "./object-select/object-select";
import { SwitchInput } from "./switch-input/switch-input";
import { AssignInputId } from "./assignId/assign-id";
import { NumberInput } from "./number-input/number-input";

export enum EditInputs {
	text = "text",
	switch = "switch",
	number = "number",
	textToggle = "text-toggle",
	select = "select",
	color = "color",
	objectSelect = "object-select",
	show = "show",
	array = "array",
	inputList = "input-list",
	conversions = "conversions",
	title = "title",
	indent = "indent",
	assign = "assign",
	assignId = "assignId",
}

type InputComponentOptions =
	| typeof InputTitle
	| typeof TextInput
	| typeof ShowInput
	| typeof ArrayInput
	| typeof InputList
	| typeof SelectInput
	| typeof ObjectSelect
	| typeof SwitchInput
	| typeof AssignInputId
	| typeof NumberInput
	| FC;

export const inputMap = new Map<EditInputs, InputComponentOptions>([
	[EditInputs.text, TextInput],
	[EditInputs.switch, SwitchInput],
	[EditInputs.number, NumberInput],
	// For toggle we can just use a wrapper around option
	// that passes a controlled value for disbled to the register option value
	[EditInputs.textToggle, () => <>text-toggle</>],
	[EditInputs.select, SelectInput],

	[EditInputs.objectSelect, ObjectSelect],

	[EditInputs.show, ShowInput],

	//////////
	[EditInputs.array, ArrayInput],
	[EditInputs.inputList, InputList],

	[EditInputs.conversions, () => <>conversions</>],

	/////////////Not an input but a title element
	[EditInputs.title, InputTitle],
	[EditInputs.indent, () => <>indent</>],

	// We can do this using an input
	// setting a default value
	// and disabling it
	[EditInputs.assign, () => <>assign</>],

	// bit of a hack fix
	// It is possible to get in a bit of a tizz with objectSelects, inputLists, and arrays
	// Where the id isn't properly passed down etc
	// So here we can force the thing - it's a workaround
	// And should probably call it objectAssign or something
	// As it is more of a sub object specifyer
	[EditInputs.assignId, AssignInputId],
]);
