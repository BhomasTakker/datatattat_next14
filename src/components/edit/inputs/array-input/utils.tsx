import { TextInputProps } from "@/types/edit/inputs/inputs";
import { InputFactory } from "../input-factory";
import { EditInputs } from "../inputs";

export const createIdentifier = (id: string) => {
	const inputOptions: TextInputProps = {
		id: `${id}.identifier`,
		type: EditInputs.text,
		label: "",
		defaultValue: "Array Item",
	};
	return <InputFactory data={inputOptions} />;
};
