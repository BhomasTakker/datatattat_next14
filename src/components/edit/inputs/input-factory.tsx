import { UnknownObject } from "@/types/utils";
import { EditInputs, inputMap } from "./inputs";

// We need to properly create types for inputs
// A whole lot of conditional types
// based on input type
type GenericInputProps = {
	type: string;
} & UnknownObject;

interface InputFactoryProps {
	data: GenericInputProps;
}

export const InputFactory = ({ data }: InputFactoryProps) => {
	const { type } = data;
	// can we pass the props type?
	const Component = inputMap.get(type as EditInputs); // Error component

	if (!Component) {
		// log warning
		return null;
	}
	// console.log("This is us", { data });
	// Need protection - incorrect type will fully blow up
	// Should perhaps pass an object?
	// @ts-expect-error - data needs typingFIX ME
	return <Component {...(data as UnknownObject)} />;
};
