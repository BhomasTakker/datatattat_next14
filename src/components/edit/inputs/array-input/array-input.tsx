import { Button } from "@/components/ui/button";
import { useContext, useState } from "react";
import { useFormContext } from "react-hook-form";
import { FaArrowDown, FaArrowUp } from "react-icons/fa6";
import { MdDelete } from "react-icons/md";

import styles from "./array-input.module.scss";
import { IconButton } from "@/components/ui/icon-button";
import { InputFactory } from "../input-factory";
import { add, move, onDelete } from "./array-input-actions";
import { randomKeyGenerator } from "@/utils/edit";
import { ArrayInputProps, GenericInput } from "@/types/edit/inputs/inputs";
import { EditContext } from "../../context/edit-context";

////////////////////////////////
// Sort types out for inputs
// generic input
type Direction = "up" | "down";

type InputWithKey = GenericInput & { key: string };
type InputList = InputWithKey[];

type InputListProps = {
	parentId: string;
	inputs: InputList;
	template: GenericInput;
	createObject: boolean;
	showControls?: boolean;
	onMove: (index: number, direction: Direction) => void;
	onDelete: (index: number) => void;
};

export const ArrayInputList = ({
	parentId,
	inputs,
	template,
	onMove,
	onDelete,
	createObject,
	showControls = true,
}: InputListProps) => {
	const { id } = template;
	return inputs.map((input, index) => {
		const inputId = createObject
			? `${parentId}.[${index}].${id}`
			: `${parentId}.[${index}]`;

		return (
			<li key={input.key} className={styles.input}>
				<InputFactory data={{ ...template, id: inputId }} />
				{showControls ? (
					<div className={styles.icons}>
						<IconButton
							data-testid="move-up"
							icon={FaArrowUp}
							onClick={() => onMove(index, "up")}
						/>
						<IconButton
							data-testid="move-down"
							icon={FaArrowDown}
							onClick={() => onMove(index, "down")}
						/>
						<IconButton
							data-testid="delete"
							icon={MdDelete}
							onClick={() => onDelete(index)}
						/>
					</div>
				) : null}
			</li>
		);
	});
};

export const ArrayInput = ({
	input,
	id,
	title,
	createObject = true,
	defaultValue = [],
	disabled = false,
}: ArrayInputProps) => {
	const { label: inputLabel, id: inputId = "" } = input;
	const { getValues, setValue, formState, handleSubmit } = useFormContext();
	const { isDirty } = formState;
	const { submitDraftHandler } = useContext(EditContext);

	const inputs: GenericInput[] = getValues(id) || defaultValue || [];

	const inputsWithUniqueKeys: InputList = inputs.map((input) => {
		return { ...input, key: randomKeyGenerator() };
	});

	const [inputList, setInputList] =
		useState<GenericInput[]>(inputsWithUniqueKeys);
	// isDirty would be better coming from the array id

	// We need to re-render the component to update the inputList
	// Whenever we move an array object, or delete
	// it should re-render the component
	// because we are updating the inputList state
	// console.log("re-rendering");
	const addInput = add({
		setValue,
		getValues,
		setInputList,
		id,
		inputId,
		createObject,
		isDirty,
		// probably something like this
	});
	const onMove = move({
		inputs: inputList,
		setValue,
		setInputList,
		id,
		isDirty,
		onDirty: handleSubmit(submitDraftHandler),
	});
	const onDeleteHnd = onDelete({
		inputs: inputList,
		setValue,
		setInputList,
		id,
		isDirty,
		onDirty: handleSubmit(submitDraftHandler),
	});

	const showControls = disabled ? false : true;

	return (
		<div className={styles.root}>
			<h2 className={styles.title}>{title}</h2>
			<ul>
				<ArrayInputList
					parentId={id}
					// This is a cheat / inputList at this point is our updated inputList
					inputs={inputList as InputList}
					template={input}
					showControls={showControls}
					onMove={onMove}
					onDelete={onDeleteHnd}
					createObject={createObject}
				/>
			</ul>

			{showControls ? (
				<Button onClick={addInput}>{`Add New ${inputLabel}`}</Button>
			) : null}
		</div>
	);
};
