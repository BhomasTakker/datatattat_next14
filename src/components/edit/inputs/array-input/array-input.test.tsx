import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { useForm, FormProvider } from "react-hook-form";
import { ArrayInput } from "./array-input";
import { EditContext } from "../../context/edit-context";
import { EditInputs } from "../inputs";

const addMock = require("./array-input-actions").add;
const moveMock = require("./array-input-actions").move;
const onDeleteMock = require("./array-input-actions").onDelete;

// This isn't correct
// We aren't rendering actual array items because we get that from form state

// Mock dependencies
jest.mock("sonner", () => ({
	toast: jest.fn(),
}));
jest.mock("./array-input-actions", () => ({
	add: jest.fn(() => jest.fn()),
	move: jest.fn(() => jest.fn()),
	onDelete: jest.fn(() => jest.fn()),
}));
jest.mock("../../../../utils/edit", () => ({
	randomKeyGenerator: jest.fn(() => Math.random().toString()),
}));
jest.mock("../input-factory", () => ({
	InputFactory: ({ data }: any) => (
		<input data-testid="input-factory" value={data.id} readOnly />
	),
}));

const defaultInput = {
	id: "test",
	label: "Test",
	type: "text",
};

const renderWithProviders = (ui: React.ReactElement, formProps: any = {}) => {
	const Wrapper: React.FC<React.PropsWithChildren<{}>> = ({ children }) => {
		const methods = useForm(formProps);
		return (
			<EditContext.Provider value={{ submitHandler: jest.fn() } as any}>
				<FormProvider {...methods}>{children}</FormProvider>
			</EditContext.Provider>
		);
	};
	return render(ui, { wrapper: Wrapper });
};

describe("ArrayInput", () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it("renders title and add button", () => {
		renderWithProviders(
			<ArrayInput
				input={defaultInput as any}
				id="arr"
				title="My Array"
				type={EditInputs.array}
			/>
		);
		expect(screen.getByText("My Array")).toBeInTheDocument();
		expect(screen.getByText("Add New Test")).toBeInTheDocument();
	});

	it("renders inputs from defaultValue", () => {
		renderWithProviders(
			<ArrayInput
				input={defaultInput as any}
				id="arr"
				title="My Array"
				// @ts-expect-error
				defaultValue={[{ foo: "bar" }, { foo: "baz" }]}
			/>
		);
		expect(screen.getAllByTestId("input-factory")).toHaveLength(2);
	});

	it("does not show controls when disabled", () => {
		renderWithProviders(
			<ArrayInput
				input={defaultInput as any}
				id="arr"
				title="My Array"
				disabled
				// @ts-expect-error
				defaultValue={[{ foo: "bar" }]}
			/>
		);
		expect(screen.queryByText("Add New Test")).not.toBeInTheDocument();
		expect(screen.queryByRole("button")).not.toBeInTheDocument();
	});

	it("calls addInput when add button is clicked", () => {
		const addHandler = jest.fn();
		addMock.mockReturnValue(addHandler);

		renderWithProviders(
			<ArrayInput
				input={defaultInput as any}
				id="arr"
				title="My Array"
				type={EditInputs.array}
			/>
		);
		fireEvent.click(screen.getByText("Add New Test"));
		expect(addHandler).toHaveBeenCalled();
	});

	it.skip("calls onMove when move up/down buttons are clicked", () => {
		const moveHandler = jest.fn();
		moveMock.mockReturnValue(moveHandler);

		renderWithProviders(
			<ArrayInput
				input={defaultInput as any}
				id="arr"
				title="My Array"
				// @ts-expect-error
				defaultValue={[{ foo: "bar" }]}
			/>
		);
		const upBtn = screen.getByTestId("move-up");
		fireEvent.click(upBtn);
		expect(moveHandler).toHaveBeenCalled();
	});

	it("calls onDelete when delete button is clicked", () => {
		const onDeleteHandler = jest.fn();
		onDeleteMock.mockReturnValue(onDeleteHandler);

		renderWithProviders(
			<ArrayInput
				input={defaultInput as any}
				id="arr"
				title="My Array"
				// @ts-expect-error
				defaultValue={[{ foo: "bar" }]}
			/>
		);
		const buttons = screen.getAllByRole("button");
		// The third button is delete (up, down, delete)
		fireEvent.click(buttons[2]);
		expect(onDeleteHandler).toHaveBeenCalled();
	});

	it("renders no inputs if defaultValue is empty", () => {
		renderWithProviders(
			<ArrayInput
				input={defaultInput as any}
				id="arr"
				title="My Array"
				defaultValue={[]}
				type={EditInputs.array}
			/>
		);
		expect(screen.queryByTestId("input-factory")).not.toBeInTheDocument();
	});

	describe("Snaps", () => {
		it("renders ArrayInput Unchanged", async () => {
			const { container } = renderWithProviders(
				<ArrayInput
					input={defaultInput as any}
					id="arr"
					title="My Array"
					defaultValue={[]}
					type={EditInputs.array}
				/>
			);
			expect(container).toMatchSnapshot();
		});
	});
});
