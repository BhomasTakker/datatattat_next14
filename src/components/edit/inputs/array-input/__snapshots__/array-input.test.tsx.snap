// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayInput Snaps renders ArrayInput Unchanged 1`] = `
<div>
  <div
    class="root"
  >
    <h2
      class="title"
    >
      My Array
    </h2>
    <ul />
    <button
      class="button "
      type="button"
    >
      Add New Test
    </button>
  </div>
</div>
`;
